src/lib/navigation/configurable-navigation-mode.ts

```diff
@@ -0,0 +1,99 @@
+
+  constructor(private navigateBackward: 'allow'|'deny'|null = null, private navigateForward: 'allow'|'deny'|null = null) {
+    super();
+    this.navigateBackward = this.navigateBackward || 'allow';
+    this.navigateForward = this.navigateForward || 'deny';
+  }
+
+  /**
+   * @inheritDoc
+   */
+  protected canTransitionToStep(wizard: WizardComponent, destinationIndex: number): boolean {
+    // if the destination step can be navigated to using the navigation bar,
+    // it should be accessible with [goToStep] as well
+    if (this.isNavigable(wizard, destinationIndex)) {
+      return true;
+    }
+
+    // navigation with [goToStep] is permitted if all previous steps to the destination step have been completed or are optional
+    return wizard.wizardSteps
+        .filter((step, index) => index < destinationIndex && index !== wizard.currentStepIndex)
+        .every(step => step.completed || step.optional);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  protected transition(wizard: WizardComponent, destinationIndex: number): void {
+    if (this.navigateForward === 'deny') {
+      // set all steps after the destination step to incomplete
+      wizard.wizardSteps
+        .filter((step, index) => wizard.currentStepIndex > destinationIndex && index > destinationIndex)
+        .forEach(step => step.completed = false);
+    }
+
+    super.transition(wizard, destinationIndex);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public isNavigable(wizard: WizardComponent, destinationIndex: number): boolean {
+    if (wizard.getStepAtIndex(destinationIndex) instanceof WizardCompletionStep) {
+      // a completion step can only be entered, if all previous steps have been completed, are optional, or selected
+      const previousStepsCompleted = wizard.wizardSteps
+        .filter((step, index) => index < destinationIndex)
+        .every(step => step.completed || step.optional || step.selected);
+      if (!previousStepsCompleted) {
+        return false;
+      }
+    }
+
+    if (destinationIndex < wizard.currentStepIndex) {
+
+      switch (this.navigateBackward) {
+        case 'allow': return true;
+        case 'deny': return false;
+      }
+
+    } else if (destinationIndex > wizard.currentStepIndex) {
+
+      switch (this.navigateForward) {
+        case 'allow': return true;
+        case 'deny': return false;
+      }
+
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * @inheritDoc
+   */
+  protected ensureCanReset(wizard: WizardComponent): void {
+    super.ensureCanReset(wizard);
+
+    if (this.navigateBackward === 'deny') {
+      // at least one step is before the default step, that is not optional
+      const illegalDefaultStep = wizard.wizardSteps
+        .filter((step, index) => index < wizard.defaultStepIndex)
+        .some(step => !step.optional);
+      if (illegalDefaultStep) {
+        throw new Error(`The default step index ${wizard.defaultStepIndex} is located after a non optional step`);
+      }
+    }

```

Comment:
There was such check in `StrictNavigationMode`, but not in `SemiStrictNavigationMode`. But these navigation modes are equal in terms of navigating back. In other words, in both of these modes the user, if the wizard is reset and there are required steps before the default step, is able to visit those required steps. I think this check really makes sense only for `"navigateBackward"="deny"`.

Reply:
The reason why the check is missing in `SemiStrictNavigationMode` is because there are basically no illegal default steps for `SemiStrictNavigationMode` except completion steps.
`SemiStrictNavigationMode` allows you to complete the wizard steps in an arbitrary order with the only condition that completion steps can only be entered when **all** other previous steps are completed.

Reply:
But why was the check present for the `StrictNavigationMode`? My point is not to bring this check to the `SemiStrictNavigationMode`, but to disable it both for `StrictNavigationMode` and `SemiStrictNavigationMode`, leaving it active only for `"navigateBackward"="deny"` (which does not correspond to any named nav modes in angular-archwizard <= 4).

Like I said, I don't think we should strictly forbid the situation when there are some required steps before the default step. It looks like bad user experience, but if the developer wants to configure the wizard this way (for whatever reason), let's let them do that. It does not break the wizard's functioning, because the user is able to navigate back from the default step to any required step before, even in strict mode (unless `"navigateBackward"="deny"`).

Reply:
The reason for the check was to help users find inconsistencies/bugs in their usage of the wizard.
Yes we can allow the use-case where the wizard initializes in a step which can only be exited in backwards direction because at least one previous step has not been completed yet, but when is such behavior intended and not considered a bug?

I think it is a better idea to introduce a new `awComplete` directive which can be used to mark wizard steps as complete during initialization. We could then change the constraint to test whether the default wizard step is located after a non-complete and non-optional step.

In addition it would allow users to define steps as completed by default.

Reply:
> The reason for the check was to help users find inconsistencies/bugs in their usage of the wizard.

This intention is good, but I think that here it goes too far. In order to hit this check, the library user needs to *explicitly* configure `[defaultStepIndex]`. In my opinion, it is safe to assume at this point that they know what they are doing...

> Yes we can allow the use-case where the wizard initializes in a step which can only be exited in backwards direction because at least one previous step has not been completed yet, but when is such behavior intended and not considered a bug?

...especially because the potential problem is so easy to discover. If the developer tries out the wizard at least once (which, in my opinion, is reasonable to assume), they will already see that the navigation is broken. And if they don't, the additional check in `reset` wouldn't help them anyway, because they would never visit the page to see the error.

> I think it is a better idea to introduce a new awComplete directive which can be used to mark wizard steps as complete during initialization. We could then change the constraint to test whether the default wizard step is located after a non-complete and non-optional step.
> 
> In addition it would allow users to define steps as completed by default.

Would not marking these steps with `awOptionalStep` work equally well in this case?

Reply:
Ok you've convinced me. Let's go with your proposal to remove the check.

> Would not marking these steps with `awOptionalStep` work equally well in this case?

No, setting a step to `optional` is different than setting it to `complete` both in their meaning and in their visualization. When defining a step as an `optional` step you tell the user that it is ok to simply skip it and/or to not complete it, while setting a step to `complete` at startup tells the user that the step is already completed with some derived/predefined information. These are two different things. I also think that we will need to add an `awComplete`/`awCompleteStep` directive long-term to solve #170 and similar issues.

Reply:
Check removed in a58b2a3. I hope you will remember to restore it (with "required" replaced with "non-complete") after `awCompleteStep` directive is implemented.

Reply:
I think you misunderstood the second part of my comment. I'm not suggesting to re-add the check with `!step.complete`. I'm just saying that I think a kind of `awCompleteStep` directive is needed to solve #170. If you have another idea how to provide the necessary functionality described in #170 I'm open for suggestions!

Reply:
Ah, okay. So we are good? Can I remove the corresponding code from the tests as well (https://github.com/madoar/angular-archwizard/pull/211#discussion_r283084170)?

Reply:
Yes, to both :)

Reply:
Removed the checks in 15271ab